### PR TITLE
CMR-5069-CMR-5070-CMR-5071

### PR DESCRIPTION
**Pull request by sxu123**
Implemented: 
CMR-5069: Multiple variables with the same name can't be associated with one collection.
CMR-5070: Each variable can only be associated with one collection.
CMR-5071: variable and collection can only be associated if they belong to the same provider.

### CODE REVIEW

File: metadata-db-app/src/cmr/metadata_db/data/memory_db.clj [hunk 1]
```diff
@@ -131,6 +132,111 @@
...
+  "Validates that collection in the concept is not associated with a different
+  variable, which has the same name as the variable in the concept.
+  Returns nil if valid and an error response if invalid."
+  [db concept]
+  (let [variable-concept-id (get-in concept [:extra-fields :variable-concept-id])
+        associated-concept-id (get-in concept [:extra-fields :associated-concept-id])]
+    (when (and variable-concept-id associated-concept-id)
+      (let [;;get variable name for the variable-concept-id from cmr_variables.
+            variable (->> @(:concepts-atom db)
+                            (filter #(= variable-concept-id (:concept-id %)))
```
Comment (daniel-zamora) [inline metadata-db-app/src/cmr/metadata_db/data/memory_db.clj]: The alignment looks off here

File: metadata-db-app/src/cmr/metadata_db/data/memory_db.clj [hunk 2]
```diff
@@ -131,6 +132,111 @@
...
+                            first)
```
Comment (daniel-zamora) [inline metadata-db-app/src/cmr/metadata_db/data/memory_db.clj]: Is it possible to have multiples of the same concept id in the collection?

File: metadata-db-app/src/cmr/metadata_db/data/memory_db.clj [hunk 3]
```diff
@@ -300,11 +406,22 @@
...
+    (and variable-concept-id associated-concept-id)))
+
 (defn save-concept
   [db provider concept]
   {:pre [(:revision-id concept)]}
 
-  (if-let [error (validate-concept-id-native-id-not-changing db provider concept)]
+  (if-let [error (or (validate-concept-id-native-id-not-changing db provider concept)
+                     (when (variable-association-concept? concept)
+                          (or (validate-same-provider-variable-association concept)
```
Comment (daniel-zamora) [inline metadata-db-app/src/cmr/metadata_db/data/memory_db.clj]: alignment seems off.

File: metadata-db-app/src/cmr/metadata_db/data/oracle/concepts.clj [hunk 4]
```diff
@@ -224,6 +224,80 @@
...
+                                  variable-concept-id associated-concept-id associated_concept_id)})))))
+
+(defn validate-same-provider-variable-association
+  "Validates that variable and the collection in the concept being saved are from the same provider.
+  Returns nil if valid and an error response if invalid."
+  [concept]
+  (let [variable-concept-id (get-in concept [:extra-fields :variable-concept-id])
+        associated-concept-id (get-in concept [:extra-fields :associated-concept-id])]
+    (when (and variable-concept-id associated-concept-id)
+      (let [v-provider-id (second (string/split variable-concept-id #"-"))
```
Comment (daniel-zamora) [inline metadata-db-app/src/cmr/metadata_db/data/oracle/concepts.clj]: Seems weird to get the provider this way.  I guess it works though.

File: metadata-db-app/src/cmr/metadata_db/data/oracle/concepts.clj [hunk 5]
```diff
@@ -224,6 +224,80 @@
...
+        associated-concept-id (get-in concept [:extra-fields :associated-concept-id])]
+    (when (and variable-concept-id associated-concept-id))
+      ;; Get all the other variables sharing the same name; not deleted
+      ;; and are associated with the same collection.
+      (let [stmt [(format "select va.variable_concept_id
+                           from   cmr_variable_associations va, 
+                                  (select concept_id, max(revision_id) maxrev
+                                   from cmr_variable_associations group by concept_id) latestva,
+                                  cmr_variables v
+                           where  va.revision_id = latestva.maxrev
```
Comment (daniel-zamora) [inline metadata-db-app/src/cmr/metadata_db/data/oracle/concepts.clj]: I personally find joins more readable when determining the relationship between tables, you can keep it this way if you like but I do think its more maintainable when you delineate the where from the join.

File: metadata-db-app/src/cmr/metadata_db/data/oracle/concepts.clj [hunk 6]
```diff
@@ -224,6 +224,80 @@
...
+  [db concept]
+  (let [variable-concept-id (get-in concept [:extra-fields :variable-concept-id])
+        associated-concept-id (get-in concept [:extra-fields :associated-concept-id])]
+    (when (and variable-concept-id associated-concept-id)
+      ;;Get all the same variables that are not deleted and are associated with other collections.
+      (let [stmt [(format "select va.associated_concept_id
+                           from   cmr_variable_associations va,
+                                  (select concept_id, max(revision_id) maxrev
+                                   from cmr_variable_associations group by concept_id) latestva
+                           where  va.revision_id = latestva.maxrev
```
Comment (daniel-zamora) [inline metadata-db-app/src/cmr/metadata_db/data/oracle/concepts.clj]: Same as above.

File: metadata-db-app/src/cmr/metadata_db/data/memory_db.clj (hunk 2 above)
Comment (sxu123) [reply metadata-db-app/src/cmr/metadata_db/data/memory_db.clj]: Yeah. at this point, it gets all the rows for all the revisions.

File: metadata-db-app/src/cmr/metadata_db/data/oracle/concepts.clj (hunk 5 above)
Comment (sxu123) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/concepts.clj]: I'm much more comfortable with the from/where.  Also, not sure how the subquery table will work with joins.  Oracle syntax seems much more strict than sybase, and not that intuitive, to me.

File: metadata-db-app/src/cmr/metadata_db/data/oracle/concepts.clj (hunk 4 above)
Comment (sxu123) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/concepts.clj]: I try not to access database.

**Comment by ygliuvt** [comment]
I couldn't find ticket numbers referenced in this PR, can you update the commit message to reference the correct ticket numbers?

**Pull request closed without merging**